src/main/java/io/vertx/ext/json/schema/openapi3/OneOfDiscriminatorValidatorFactory.java

```diff
@@ -0,0 +1,113 @@
+
+import static io.vertx.ext.json.schema.ValidationException.createException;
+import java.util.List;
+import java.util.Map;
+
+public class OneOfDiscriminatorValidatorFactory extends BaseCombinatorsValidatorFactory {
+
+    @Override
+    public Validator createValidator(JsonObject schema, JsonPointer scope, SchemaParserInternal parser, MutableStateValidator parent) {
+      OneOfValidator validator = (OneOfValidator) super.createValidator(schema, scope, parser, parent);
+      if (schema.containsKey("discriminator") && schema.getValue("discriminator") instanceof JsonObject) {
+        JsonObject discriminator = schema.getJsonObject("discriminator");
+        validator.setPropertyName(discriminator.getString("propertyName"));
+        if (discriminator.containsKey("mapping")) {
+          validator.setMapping(discriminator.getJsonObject("mapping").getMap().entrySet().stream()
+                    .collect(Collectors.toMap(entry -> entry.getKey(), entry -> entry.getValue().toString()))
+                );
+        }
+      }
+      return validator;
+    }
+
+  @Override
+  protected BaseCombinatorsValidator instantiate(MutableStateValidator parent) {
+    return new OneOfValidator(parent);
+  }
+
+  @Override
+  protected String getKeyword() {
+    return "oneOf";
+  }
+
+  class OneOfValidator extends BaseCombinatorsValidator {
+
+    private String propertyName;
+    private Map<String, String> mapping;
+
+    public OneOfValidator(MutableStateValidator parent) {
+      super(parent);
+    }
+
+    private boolean isValidSync(SchemaInternal schema, ValidatorContext context, Object in) {
+      try {
+        schema.validateSync(context, in);
+        return true;
+      } catch (ValidationException e) {
+        return false;
+      }
+    }
+
+    @Override
+    public void validateSync(ValidatorContext context, Object in) throws ValidationException, NoSyncValidationException {
+      this.checkSync();
+      long validCount = Arrays.stream(schemas).filter(s -> checkDiscriminator(s, in)).map(s -> isValidSync(s, context, in)).filter(b -> b.equals(true)).count();
+      if (validCount > 1) throw createException("More than one schema valid", "oneOf", in);
+      else if (validCount == 0) throw createException("No schema matches", "oneOf", in);
+    }
+
+    @Override
+    public Future<Void> validateAsync(ValidatorContext context, Object in) {
+      if (isSync()) return validateSyncAsAsync(context, in);
+      List<Future<Void>> futures = Arrays.stream(schemas).filter(s -> checkDiscriminator(s, in)).map(s -> s.validateAsync(context, in)).collect(Collectors.toList());
+      if (futures.isEmpty()) {
+        return Future.failedFuture(createException("No schema matches", "oneOf", in));
+      } else {
+        return FutureUtils.oneOf(futures).recover(err -> Future.failedFuture(createException("No schema matches", "oneOf", in, err)));
+      }
+    }
+
+    private boolean checkDiscriminator(SchemaInternal schema, Object in) {
```

Comment:
I'm wondering, can we place some of this code that parses the original schema in the factory/initialization code more than on the hot path?

You can also reorganize the data structure of the validator to have a map discriminator -> schema. Note that you don't have to use the same validator both for oneOf case and oneOf + discriminator case. Feel free to reach me if you have more questions

Reply:
ping @dps123

Reply:
@slinkydeveloper Moved the schema parsing to the validator initializer

Reply:
Can you please check my comments on https://github.com/eclipse-vertx/vertx-json-schema/issues/31?